### PR TITLE
`ultralytics 8.4.49` Fail fast on three common training failure modes

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Three small fixes to surface failure modes seen at scale in worker logs before they consume GPU time or produce cryptic errors:

- **`torch_safe_load`** — detect corrupt cached `.pt` (`PytorchStreamReader` / zip central directory error), delete the local file, and re-fetch via the existing `safe_download` retry path. A truncated cache previously failed every subsequent run for that file with no recovery.
- **`BaseTrainer._setup_train`** — after the freeze loop, raise a clear `RuntimeError` if no parameters have `requires_grad`. A too-large `freeze` value (e.g. `freeze=100`) previously crashed mid-step with `element 0 of tensors does not require grad and does not have a grad_fn`.
- **`ClassificationTrainer.get_dataloader`** — raise `FileNotFoundError` naming the empty split if the cls dataset has zero samples after class filtering. Previously `RandomSampler` raised `num_samples should be a positive integer value, but got num_samples=0` with no hint about the dataset.

Each fix is local, additive, and mirrors existing patterns in the same function — no new abstractions, no duplicated retry logic (corrupt-cache reload reuses `attempt_download_asset` → `safe_download`).

Also bumps version to `8.4.49`.

## Test plan
- [ ] `pytest tests/` passes
- [ ] Train with `freeze=999` raises the new `RuntimeError` before AMP check
- [ ] Train cls on a dataset with empty `train/` raises the new `FileNotFoundError` before AutoBatch
- [ ] Truncate a cached `.pt`, run training: re-download fires, training proceeds

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ This PR improves training and model loading reliability by adding clearer error checks for frozen models, empty classification datasets, mismatched classes, and corrupted cached weight files.

### 📊 Key Changes
- ✅ Added a training safeguard in `trainer.py` that stops execution if `freeze=` disables **all** model parameters, leaving nothing trainable.
- 🖼️ Added a classification dataset check in `classify/train.py` to raise a clear `FileNotFoundError` when no images are found in a dataset split.
- 🏷️ Improved class-mismatch handling for classification training:
  - Samples with invalid class indices are still filtered out.
  - A new error is now raised if **all** samples get filtered, preventing confusing downstream failures.
- 📦 Refactored weight loading in `tasks.py` into a reusable internal loader function.
- 🔄 Added automatic recovery for corrupted downloaded `.pt` weight caches:
  - If a cached remote weight file is truncated or broken, it is deleted and re-downloaded automatically.
  - Local user-supplied files are not deleted, avoiding accidental data loss.
- 🔢 Bumped the package version from `8.4.48` to `8.4.49`.

### 🎯 Purpose & Impact
- 🚫 Prevents wasted training runs by catching invalid `freeze` settings early with a clear message.
- 🧭 Makes classification setup easier to debug by reporting missing images and dataset formatting issues sooner.
- 🧹 Avoids obscure CUDA or dataloader errors when dataset class labels do not match the model’s expected class count.
- ♻️ Improves robustness when loading pretrained models by automatically fixing corrupted cached downloads.
- 🔒 Reduces risk for users by only auto-deleting broken **cached remote files**, not local checkpoint files.
- 😊 Overall, this PR makes Ultralytics training and checkpoint loading more reliable, with clearer feedback and fewer hard-to-diagnose failures.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting this `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist shortly. Please review the checklist below for a smooth integration ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/tasks.py [hunk 1]
```diff
@@ -1462,6 +1462,15 @@ def torch_safe_load(weight, safe_only=False):
             else:
                 ckpt = torch_load(file, map_location="cpu")
 
+    except RuntimeError as e:
+        # Corrupt cached file (e.g. truncated download); delete and re-fetch via existing safe_download retry
+        if "PytorchStreamReader" not in str(e):
+            raise
+        LOGGER.warning(f"Corrupt cache {file}, re-downloading {weight}...")
+        Path(file).unlink(missing_ok=True)
+        file = attempt_download_asset(weight)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:1474]: ⚠️ **HIGH**: This recovery path will also run for a user-supplied local checkpoint. In that case we delete the local `.pt` and then hand the same path to `attempt_download_asset()`, which turns a local corruption into a destructive side effect plus a misleading download attempt. Please gate the unlink/re-download logic to cached/downloaded assets only, or explicitly skip it when `weight` already exists as a local file.

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -1462,6 +1462,15 @@ def torch_safe_load(weight, safe_only=False):
...
+        ckpt = torch_load(file, map_location="cpu")
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py]: ⚠️ **HIGH**: The retry ignores `safe_only` and always uses `torch_load(...)`. If `torch_safe_load(..., safe_only=True)` hits this branch, the second load silently drops the SafeUnpickler path the caller requested. The retry should mirror the original branch so `safe_only` is preserved.

File: ultralytics/models/yolo/classify/train.py [hunk 3]
```diff
@@ -151,6 +151,11 @@ def get_dataloader(self, dataset_path: str, batch_size: int = 16, rank: int = 0,
                 f"Skipping {skipped} samples from extra classes: {extra_classes}"
             )
 
+        if not dataset.samples:
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/classify/train.py]: 💡 **MEDIUM**: `dataset.samples` can be empty even when the split contains images, for example after class filtering removes every sample. In that case this raises `FileNotFoundError` with a dataset-format hint, which is misleading. Please distinguish between 'no images discovered on disk' and 'all samples filtered out' so users get the right remediation.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Solid fail-fast improvements overall. The main issues are in `torch_safe_load()`: the new recovery path can delete user-local checkpoints and it drops the `safe_only` guarantee on retry. The classification check is useful, but its current error message can misdiagnose fully filtered splits.

💬 Posted 3 inline comments

**Comment by glenn-jocher** [comment]
Addressed the three review findings in 73ef6b2b5:

- **`torch_safe_load` destructive unlink (HIGH)** — the corrupt-cache recovery path is now gated to URL inputs (`http://`/`https://`). User-supplied local `.pt` files surface the original `PytorchStreamReader` error instead of getting deleted.
- **`safe_only` dropped on retry (HIGH)** — extracted the load body into a local `_load()` helper used by both the initial attempt and the retry, so `safe_only=True` keeps the `SafeUnpickler` path on the second load.
- **cls split error misdiagnosis (MEDIUM)** — the existing `FileNotFoundError("No images found...")` now fires *before* the `nc` filter, so it only applies to genuinely empty splits. A separate `RuntimeError` covers the all-filtered case with a clear "model nc vs dataset class indices" hint.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Thanks for this thoughtful reliability upgrade, @glenn-jocher.

“As we may think, so we become.” — Buddha

That fits this PR well: by making failure cases clearer and recovery smarter, you’re helping users get faster feedback, avoid wasted training runs, and trust model loading more across the board. The added safeguards for fully frozen models, empty classification splits, class mismatches, and corrupted cached weights make Ultralytics training more robust and much easier to debug.

Really appreciate the care here—small checks, big impact.
